astropy/table/table.py

```diff
@@ -2918,8 +2918,12 @@ def sort(self, keys=None, reverse=False):
             prev_frozen = sort_index._frozen
             sort_index._frozen = True
 
-        for col in self.columns.values():
-            col[:] = col.take(indexes, axis=0)
+        for name, col in self.columns.items():
+            new_col = col.take(indexes, axis=0)
+            try:
+                col[:] = new_col
+            except Exception:
+                self[col.info.name] = new_col
```

Comment:
I tried setting to the new column in all cases but this caused problems with indexes which would require a deeper dive to understand / fix.

Reply:
OK - I think here having a solution is definitely much better than worry about perfection. Maybe a new issue, though?